##MOUNTED DRIVE

In [83]:
from google.colab import drive
drive.mount('/content/gdrive')
import os
os.chdir("/content/gdrive/My Drive/HW1_ML")
import numpy as np
from util import *
from logistic_np import *

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


## LOGISTIC_NP

###TODO 1.1 & 1.2: Chuẩn hóa dữ liệu

In [84]:
train_x, train_y, test_x, test_y = get_vehicle_data()
train_x, test_x = normalize_per_pixel(train_x, test_x)
train_x[0]

EOF Reached
Done reading


array([[-0.75118653, -1.15748498, -1.04647363, ..., -0.75180351,
         1.42435733,  0.59829548],
       [-1.18627331, -0.79803185, -0.61205517, ...,  0.04259907,
         1.38566852,  0.40350883],
       [ 0.09870111, -0.47991064, -1.08492691, ...,  0.75008153,
        -0.07413474,  0.28723528],
       ...,
       [-0.56416359,  0.63954484,  1.44473651, ..., -0.32600671,
        -0.25256551, -0.18010676],
       [-0.7649102 , -0.17853185,  1.14028905, ..., -0.28271976,
        -0.22505078, -0.18389908],
       [ 0.10841937, -0.15449722,  0.35621025, ..., -0.24225544,
        -0.18119004, -0.15247033]])

In [85]:
train_x, train_y, test_x, test_y = get_vehicle_data()
train_x, test_x= normalize_all_pixel(train_x, test_x)
train_x[0]

EOF Reached
Done reading


array([[-0.72765297, -1.08908313, -1.00302833, ..., -0.76781188,
         1.23439647,  0.47711423],
       [-1.10629409, -0.76207489, -0.62438721, ..., -0.03921457,
         1.18276359,  0.28205668],
       [ 0.0181553 , -0.4981735 , -1.04892422, ...,  0.60332794,
        -0.14821732,  0.16731695],
       ...,
       [-0.60717625,  0.50006218,  1.24587044, ..., -0.43506665,
        -0.34901185, -0.28016801],
       [-0.79075982, -0.24574609,  0.98196906, ..., -0.38917076,
        -0.32032692, -0.26869404],
       [ 0.02962927, -0.22279814,  0.25337175, ..., -0.33753788,
        -0.26869404, -0.23427212]])

###TODO 1.3: Reshape dữ liệu 

In [86]:
train_x = reshape2D(train_x)

In [87]:
train_x.shape

(2400, 4096)

###TODO 1.4: Thêm đặc trưng

In [88]:
train_x = add_one(train_x)

In [89]:
train_x.shape

(2400, 4097)

###TODO 1.5: Tính các giá trị phân loại 

In [90]:
num_feature = train_x.shape[1]
bin_classifier = LogisticClassifier((num_feature, 1))
y_hat = bin_classifier.feed_forward(train_x)

In [91]:
y_hat

array([[0.61083233],
       [0.68806083],
       [0.83222221],
       ...,
       [0.25894807],
       [0.79640414],
       [0.51988554]])

###TODO 1.6: Tính độ lỗi 

In [92]:
loss = bin_classifier.compute_loss(train_y, y_hat)

In [93]:
loss

0.811250412816247

###TODO 1.7: Tính đạo hàm 

In [94]:
grad = bin_classifier.get_grad(train_x, train_y, y_hat)

In [95]:
grad

array([[ 0.01630328],
       [ 0.00905221],
       [-0.00020457],
       ...,
       [-0.0046142 ],
       [-0.00787064],
       [ 0.03974735]])

###TODO 1.8: Cập nhật w


In [96]:
bin_classifier.update_weight(grad, learning_rate=0.01)

In [97]:
bin_classifier.w 

array([[-0.0062773 ],
       [ 0.01276355],
       [ 0.04746383],
       ...,
       [-0.00823311],
       [-0.02151544],
       [-0.02710052]])

###TODO 1.9: Cập nhật w với momentum

In [98]:
bin_classifier = LogisticClassifier((num_feature, 1))

In [99]:
momentum = np.ones_like(grad)
learning_rate = 0.01
momentum_rate = 0.9

In [100]:
bin_classifier.update_weight_momentum(grad, learning_rate, momentum, momentum_rate)

In [101]:
bin_classifier.w

array([[-0.90200168],
       [-0.90677683],
       [-0.89980481],
       ...,
       [-0.89688172],
       [-0.92481079],
       [-0.86835384]])

###TODO 1.10: Đánh giá mô hình phân loại

In [102]:
test_x = reshape2D(test_x)
test_x = add_one(test_x)

In [103]:
y_hat = bin_classifier.feed_forward(test_x)

In [104]:
test(y_hat, test_y)

Precision: 0.713
Recall: 0.703
F1-score: 0.708


(0.7128378378378378, 0.7033333333333334, 0.7080536912751677)

##TEST LOGISTIC_NP

In [105]:
!python3 logistic_np.py

EOF Reached
Done reading
Precision: 0.751
Recall: 0.877
F1-score: 0.809


In [106]:
!python3 unit_test.py

Please enter 0 for logistic or 1 for softmax: 0
Running logistic np unit test...
EOF Reached
Done reading
Testing normalize_per_pixel: Passed.
----------------------------------------
EOF Reached
Done reading
Testing normalize_all_pixel: Passed.
----------------------------------------
Testing reshape2D: Passed.
----------------------------------------
Testing add_one: Passed.
----------------------------------------
Testing feed_forward 1: Passed.
----------------------------------------
Testing compute_loss 1: Passed.
----------------------------------------
Testing get_grad 1: Passed.
----------------------------------------
Testing update_weight 1: Passed.
----------------------------------------
Testing update_weight_momentum 1: Passed.
----------------------------------------
Testing feed_forward 2: Passed.
----------------------------------------
Testing compute_loss 2: Passed.
----------------------------------------
Testing get_grad 2: Passed.
---------------------------------

##LOGISTIC_TF

###Chuẩn hóa dữ liệu

In [107]:
import tensorflow as tf
np.random.seed(2018)
tf.compat.v1.set_random_seed(2018)

train_x, train_y, test_x, test_y = get_vehicle_data()
train_x, test_x = normalize_per_pixel(train_x, test_x)
train_x = reshape2D(train_x)
test_x = reshape2D(test_x)
train_x = add_one(train_x)
test_x = add_one(test_x)

m, n = train_x.shape
num_epoch = 100
tf.compat.v1.disable_eager_execution()
tf.compat.v1.disable_v2_behavior()

EOF Reached
Done reading


###TODO 1.11: Tạo TF place holders 

In [108]:
x = tf.compat.v1.placeholder(tf.float32, shape=(None, n))
y = tf.compat.v1.placeholder(tf.float32, shape=(None, 1))

###TODO 1.12: Tạo TF variable cho w

In [109]:
w = tf.Variable(tf.ones([n, 1]), dtype=tf.float32)

###TODO 1.13 & TODO 1.14: Tạo toán tử feed forward & Tính độ lỗi 

In [110]:
pred = tf.matmul(x, w)
cost = tf.nn.sigmoid_cross_entropy_with_logits(logits = pred, labels = y)

###TODO 1.15: Tạo đối tượng SGD optimizer 

In [111]:
SGD = tf.compat.v1.train.GradientDescentOptimizer(learning_rate=0.002).minimize(cost)

###TODO 1.16: Tính cost và cập nhật w

In [112]:
init = tf.compat.v1.global_variables_initializer()
with tf.compat.v1.Session() as sess:
    sess.run(init)
    for e in range(num_epoch):
        sess.run(SGD, feed_dict={x: train_x, y: train_y})
        cal_cost = sess.run(cost, feed_dict={x: train_x, y: train_y})
        Weight = sess.run(w)  
    y_hat = sess.run(pred, feed_dict={x: test_x})
    test(y_hat, test_y)

Precision: 0.759
Recall: 0.840
F1-score: 0.797
